server/schedulers/balance_leader.go

```diff
@@ -137,6 +144,58 @@ func (l *balanceLeaderScheduler) IsScheduleAllowed(cluster schedule.Cluster) boo
 	return allowed
 }
 
+type candidateStores struct {
+	stores        []*core.StoreInfo
+	storeIndexMap map[uint64]int
+	index         int
+	compareOption func([]*core.StoreInfo) func(int, int) bool
+}
+
+func NewCandidateStores(stores []*core.StoreInfo, compareOption func([]*core.StoreInfo) func(int, int) bool) *candidateStores {
+	cs := &candidateStores{stores: stores, compareOption: compareOption}
+	cs.storeIndexMap = map[uint64]int{}
+	cs.initSort()
+	return cs
+}
+
+// hasStore returns returns true when there are leftover stores.
+func (cs *candidateStores) hasStore() bool {
+	return cs.index < len(cs.stores)
+}
+
+func (cs *candidateStores) getStore() *core.StoreInfo {
+	return cs.stores[cs.index]
+}
+
+func (cs *candidateStores) next() {
+	cs.index++
+}
+
+func (cs *candidateStores) initSort() {
+	sort.Slice(cs.stores, cs.compareOption(cs.stores))
+	for i := 0; i < len(cs.stores); i++ {
+		cs.storeIndexMap[cs.stores[i].GetID()] = i
+	}
+}
+
+func (cs *candidateStores) reSort(stores ...*core.StoreInfo) {
+	if !cs.hasStore() {
+		return
+	}
+	for _, store := range stores {
+		index, ok := cs.storeIndexMap[store.GetID()]
+		if !ok {
+			continue
+		}
+		resortStores(cs.stores, cs.storeIndexMap, index, cs.compareOption(cs.stores))
+	}
+}
+
+func leaderScore(store *core.StoreInfo, plan *balancePlan) float64 {
```

Comment:
I prefer to remove this function and move GetOpInfluence directly into `xxxOption`.